plugin/static_account.go

```diff
@@ -0,0 +1,344 @@
+	oldBindings := a.Bindings
+	bindsToAdd := newBindings.sub(oldBindings)
+	bindsToRemove := oldBindings.sub(newBindings)
+
+	b.Logger().Debug("updating bindings for static account")
+	walIds, err := b.addWalsForStaticAccountBindings(ctx, req, a, bindsToAdd, bindsToRemove)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := b.createIamBindings(ctx, req, a.EmailOrId, bindsToAdd); err != nil {
+		return nil, err
+	}
+
+	if err := b.removeBindings(ctx, req, a.EmailOrId, bindsToRemove); err != nil {
+		return nil, err
+	}
+
+	return walIds, nil
+}
+
+// addWalsForStaticAccountResources creates WALs to clean up a roleset's service account, bindings, and a key if needed.
+func (b *backend) addWalsForStaticAccountResources(ctx context.Context, req *logical.Request, staticAcctName string, boundResources *gcpAccountResources) (walIds []string, err error) {
+	if boundResources == nil {
+		b.Logger().Debug("skip WALs for nil GCP account resources")
+		return nil, nil
+	}
+
+	walIds = make([]string, 0, len(boundResources.bindings)+1)
+	for resName, roles := range boundResources.bindings {
+		walId, err := framework.PutWAL(ctx, req.Storage, walTypeIamPolicyDiff, &walIamPolicyStaticAccount{
+			StaticAccount: staticAcctName,
+			AccountId:     boundResources.accountId,
+			Resource:      resName,
+			RolesAdded:    roles.ToSlice(),
+		})
+		if err != nil {
+			return walIds, errwrap.Wrapf("unable to create WAL entry to clean up service account bindings: {{err}}", err)
+		}
+		walIds = append(walIds, walId)
+	}
+
+	if boundResources.tokenGen != nil {
+		walId, err := b.addWalStaticAccountServiceAccountKey(ctx, req, staticAcctName, &boundResources.accountId, boundResources.tokenGen.KeyName)
+		if err != nil {
+			return walIds, err
+		}
+		walIds = append(walIds, walId)
+	}
+	return walIds, nil
+}
+
+func (b *backend) addWalsForStaticAccountBindings(ctx context.Context, req *logical.Request, a *StaticAccount, removed, added ResourceBindings) (walIds []string, err error) {
+	walIds = make([]string, 0, len(removed)+len(added))
+
+	// Add WALs for resources in added bindings
+	for resource, rolesAdded := range added {
+		walEntry := &walIamPolicyStaticAccount{
+			AccountId:  a.ServiceAccountId,
+			Resource:   resource,
+			RolesAdded: rolesAdded.ToSlice(),
+		}
+		if rolesRemoved, ok := removed[resource]; ok {
+			walEntry.RolesRemoved = rolesRemoved.ToSlice()
+		}
+		walId, err := framework.PutWAL(ctx, req.Storage, walTypeIamPolicyDiff, walEntry)
+		if err != nil {
+			return walIds, errwrap.Wrapf("unable to create WAL entry to clean up service account bindings: {{err}}", err)
+		}
+		walIds = append(walIds, walId)
+	}
+
+	// Add WALs for resources in removed bindings not in added bindings.
+	for resource, rolesRemoved := range removed {
+		if _, ok := added[resource]; ok {
+			continue
+		}
+		walEntry := &walIamPolicyStaticAccount{
+			AccountId:    a.ServiceAccountId,
```

Comment:
Similar to comment above, I think we'll want to add `StaticAccount: a.Name` to this WAL entry.